src/main/java/zd/zero/waifu/motivator/plugin/assets/RemoteAssetDefinitionService.kt

```diff
@@ -1,16 +1,34 @@
 package zd.zero.waifu.motivator.plugin.assets
 
+import zd.zero.waifu.motivator.plugin.tools.toOptional
+import java.util.*
 import kotlin.random.Random
 
-abstract class RemoteAssetDefinitionService<T : AssetDefinition>(
-    private val remoteAssetManager: RemoteAssetManager<T>
+abstract class RemoteAssetDefinitionService<T : AssetDefinition, U : Asset>(
+    private val remoteAssetManager: RemoteAssetManager<T, U>
 ) {
     private val random = Random(System.currentTimeMillis())
 
-    fun getRandomAssetByCategory(waifuAssetCategory: WaifuAssetCategory): T =
-        remoteAssetManager.resolveAsset(
-            remoteAssetManager.supplyAssetDefinitions()
-                .filter { it.categories.contains(waifuAssetCategory) }
-                .random(random)
-        )
+    fun getRandomAssetByCategory(waifuAssetCategory: WaifuAssetCategory): Optional<U> =
+        pickRandomAsset(remoteAssetManager.supplyAssetDefinitions(), waifuAssetCategory)
```

Comment:
This is the core functionality right here.

The plugin will always try to fetch a asset from the complete list. (They may come back online while still coding, so it's worth always trying and not keeping track of network status)
Which means that it could not be downloaded yet, that way the user can get new stuff when they need it.

However, if they happen to be offline, the plugin will fall back to any local assets of the same category.
If there isn't any local assets that don't match, well you don't get anything. 

Which means that code that calls this, can just try again for a different category and hope it gets something.